system/lib/wasmfs/proxied_async_js_impl_backend.h

```diff
@@ -0,0 +1,192 @@
+
+void _wasmfs_jsimpl_async_alloc_file(js_index_t backend,
+                                     js_index_t index,
+                                     async_callback_t callback,
+                                     CppCallbackState* state);
+void _wasmfs_jsimpl_async_free_file(js_index_t backend,
+                                    js_index_t index,
+                                    async_callback_t callback,
+                                    CppCallbackState* state);
+void _wasmfs_jsimpl_async_write(js_index_t backend,
+                                js_index_t index,
+                                const uint8_t* buffer,
+                                size_t length,
+                                off_t offset,
+                                async_callback_t callback,
+                                CppCallbackState* state);
+void _wasmfs_jsimpl_async_read(js_index_t backend,
+                               js_index_t index,
+                               const uint8_t* buffer,
+                               size_t length,
+                               off_t offset,
+                               async_callback_t callback,
+                               CppCallbackState* state);
+void _wasmfs_jsimpl_async_get_size(js_index_t backend,
+                                   js_index_t index,
+                                   async_callback_t callback,
+                                   CppCallbackState* state);
+}
+
+namespace wasmfs {
+
+class ProxiedAsyncJSImplFile : public DataFile {
+  emscripten::SyncToAsync& proxy;
+
+  js_index_t getBackendIndex() {
+    static_assert(sizeof(backend_t) == sizeof(js_index_t), "TODO: wasm64");
+    return js_index_t(getBackend());
+  }
+
+  js_index_t getFileIndex() {
+    static_assert(sizeof(this) == sizeof(js_index_t), "TODO: wasm64");
+    return js_index_t(this);
+  }
+
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+    CppCallbackState state;
+
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      state.resume = resume;
+
+      _wasmfs_jsimpl_async_write(
+        getBackendIndex(),
+        getFileIndex(),
+        buf,
+        len,
+        offset,
+        [](CppCallbackState* state) { (*state->resume)(); },
+        &state);
+    });
+
+    return state.result;
+  }
+
+  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
+    // The caller should have already checked that the offset + len does
+    // not exceed the file's size.
+    assert(offset + len <= getSize());
+
+    CppCallbackState state;
+
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      state.resume = resume;
+
+      _wasmfs_jsimpl_async_read(
+        getBackendIndex(),
+        getFileIndex(),
+        buf,
+        len,
+        offset,
+        [](CppCallbackState* state) { (*state->resume)(); },
+        &state);
+    });
+
+    return state.result;
+  }
+
+  void flush() override {}
```

Comment:
Should we abort with some sort of "unimplemented" message here?

Reply:
This matches `js_impl_backend`, but fair point, maybe we should do the same in both. OTOH, I think a default of "sync does nothing" is reasonable. I guess we should decide between those eventually, and if we do go with the latter, we should probably put the default in File and not here and in JSImpl.

Reply:
Ok, leaving this for now seems fine.